File: backend/src/services/user_service.py
"""用户服务：管理用户数据"""
from typing import Optional, List, Tuple
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from sqlalchemy import desc
from ..database import SessionLocal
from ..db_models import UserModel
from ..models import User


class UserService:
    """用户服务类"""
    
    def __init__(self):
        """初始化用户服务"""
        pass
    
    def _get_db(self):
        """获取数据库会话（生成器）"""
        db = SessionLocal()
        try:
            yield db
        finally:
            db.close()
    
    def create_user(self, username: str, password: str, nickname: Optional[str] = None, email: Optional[str] = None, phone: Optional[str] = None, avatar: Optional[str] = None) -> User:
        """
        创建新用户
        
        Args:
            username: 用户名（必填，必须唯一）
            password: 用户密码（明文）
            nickname: 用户昵称（可选，用于显示）
            email: 用户邮箱（可选，用于登录）
            phone: 用户手机号（可选，用于登录）
            avatar: 用户头像URL（可选）
            
        Returns:
            User: 创建的用户实体
            
        Raises:
            ValueError: 用户名、邮箱或手机号已存在
        """
        # 使用User.create方法创建用户实体（密码自动加密）
        user_entity = User.create(username=username, password=password, nickname=nickname, email=email, phone=phone, avatar=avatar)
        
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            # 转换为SQLAlchemy模型
            user_model = UserModel(
                user_id=user_entity.user_id,
                username=user_entity.username,
                nickname=user_entity.nickname,
                email=user_entity.email,
                phone=user_entity.phone,
                password_hash=user_entity.password_hash,
                avatar=user_entity.avatar,
                created_at=user_entity.created_at
            )
            
            db.add(user_model)
            db.commit()
            db.refresh(user_model)
            
            # 转换回User实体
            return User(
                user_id=user_model.user_id,
                username=user_model.username,
                nickname=user_model.nickname,
                email=user_model.email,
                phone=user_model.phone,
                password_hash=user_model.password_hash,
                avatar=user_model.avatar,
                created_at=user_model.created_at
            )
        except IntegrityError as e:
            db.rollback()
            # 判断是用户名、邮箱还是手机号冲突
            if db.query(UserModel).filter(UserModel.username == username).first():
                raise ValueError("用户名已存在")
            if email and db.query(UserModel).filter(UserModel.email == email).first():
                raise ValueError("邮箱已存在")
            if phone and db.query(UserModel).filter(UserModel.phone == phone).first():
                raise ValueError("手机号已存在")
            raise ValueError("创建用户失败")
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass
    
    def get_user_by_email(self, email: str) -> Optional[User]:
        """
        根据邮箱获取用户
        
        Args:
            email: 用户邮箱
            
        Returns:
            User: 用户实体，如果不存在返回None
        """
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            user_model = db.query(UserModel).filter(UserModel.email == email).first()
            if user_model is None:
                return None
            
            return User(
                user_id=user_model.user_id,
                username=user_model.username,
                nickname=user_model.nickname,
                email=user_model.email,
                phone=user_model.phone,
                password_hash=user_model.password_hash,
                avatar=user_model.avatar,
                created_at=user_model.created_at
            )
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass
    
    def get_user_by_phone(self, phone: str) -> Optional[User]:
        """
        根据手机号获取用户
        
        Args:
            phone: 用户手机号
            
        Returns:
            User: 用户实体，如果不存在返回None
        """
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            user_model = db.query(UserModel).filter(UserModel.phone == phone).first()
            if user_model is None:
                return None
            
            return User(
                user_id=user_model.user_id,
                username=user_model.username,
                nickname=user_model.nickname,
                email=user_model.email,
                phone=user_model.phone,
                password_hash=user_model.password_hash,
                avatar=user_model.avatar,
                created_at=user_model.created_at
            )
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass
    
    def get_user_by_username(self, username: str) -> Optional[User]:
        """
        根据用户名获取用户
        
        Args:
            username: 用户名
            
        Returns:
            User: 用户实体，如果不存在返回None
        """
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            user_model = db.query(UserModel).filter(UserModel.username == username).first()
            if user_model is None:
                return None
            
            return User(
                user_id=user_model.user_id,
                username=user_model.username,
                nickname=user_model.nickname,
                email=user_model.email,
                phone=user_model.phone,
                password_hash=user_model.password_hash,
                avatar=user_model.avatar,
                created_at=user_model.created_at
            )
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass
    
    def get_user_by_id(self, user_id: str) -> Optional[User]:
        """
        根据用户ID获取用户
        
        Args:
            user_id: 用户ID（UUID）
            
        Returns:
            User: 用户实体，如果不存在返回None
        """
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            user_model = db.query(UserModel).filter(UserModel.user_id == user_id).first()
            if user_model is None:
                return None
            
            return User(
                user_id=user_model.user_id,
                username=user_model.username,
                nickname=user_model.nickname,
                email=user_model.email,
                phone=user_model.phone,
                password_hash=user_model.password_hash,
                avatar=user_model.avatar,
                created_at=user_model.created_at
            )
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass
    
    def get_all_users(self, page: int = 1, page_size: int = 20) -> Tuple[List[User], int]:
        """
        获取所有用户列表（支持分页）
        
        Args:
            page: 页码（从1开始）
            page_size: 每页数量
            
        Returns:
            Tuple[List[User], int]: (用户列表, 总数)
        """
        db_gen = self._get_db()
        db = next(db_gen)
        try:
            # 计算偏移量
            offset = (page - 1) * page_size
            
            # 获取总数
            total = db.query(UserModel).count()
            
            # 获取分页数据（按创建时间倒序）
            user_models = db.query(UserModel).order_by(desc(UserModel.created_at)).offset(offset).limit(page_size).all()
            
            # 转换为User实体列表
            users = [
                User(
                    user_id=user_model.user_id,
                    username=user_model.username,
                    nickname=user_model.nickname,
                    email=user_model.email,
                    phone=user_model.phone,
                    password_hash=user_model.password_hash,
                    avatar=user_model.avatar,
                    created_at=user_model.created_at
                )
                for user_model in user_models
            ]
            
            return users, total
        finally:
            try:
                next(db_gen, None)  # 完成生成器
            except StopIteration:
                pass

File: backend-java/src/main/java/com/aieducenter/studio/application/service/UserService.java
package com.platform.application.service;

import com.platform.domain.user.User;
import com.platform.infrastructure.persistence.jpa.UserEntity;
import com.platform.infrastructure.persistence.jpa.UserJpaRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * 用户服务
 *
 * 对应Python: backend/src/services/user_service.py
 * 负责用户的CRUD操作
 *
 * @author AI Teacher Platform
 */
@Service
public class UserService {

    private static final Logger logger = LoggerFactory.getLogger(UserService.class);

    private final UserJpaRepository userJpaRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(
            UserJpaRepository userJpaRepository,
            PasswordEncoder passwordEncoder
    ) {
        this.userJpaRepository = userJpaRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * 创建新用户
     *
     * 对应Python: def create_user(...)
     *
     * @param username 用户名
     * @param password 密码
     * @param nickname 昵称
     * @param email 邮箱
     * @param phone 手机号
     * @param avatar 头像
     * @param isAdmin 是否为管理员
     * @return 创建的用户
     * @throws IllegalArgumentException 用户名/邮箱/手机号已存在
     */
    @Transactional
    public User createUser(
            String username,
            String password,
            String nickname,
            String email,
            String phone,
            String avatar,
            Boolean isAdmin
    ) {
        // 检查用户名是否已存在
        if (userJpaRepository.existsByUsername(username)) {
            throw new IllegalArgumentException("用户名已存在: " + username);
        }

        // 检查邮箱是否已存在
        if (email != null && !email.trim().isEmpty() && userJpaRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("邮箱已存在: " + email);
        }

        // 检查手机号是否已存在
        if (phone != null && !phone.trim().isEmpty() && userJpaRepository.existsByPhone(phone)) {
            throw new IllegalArgumentException("手机号已存在: " + phone);
        }

        // 创建用户实体
        UserEntity entity = new UserEntity();
        entity.setUserId(UUID.randomUUID().toString());
        entity.setUsername(username);
        entity.setPasswordHash(passwordEncoder.encode(password));
        entity.setNickname(nickname);
        entity.setEmail(email);
        entity.setPhone(phone);
        entity.setAvatar(avatar);
        entity.setIsAdmin(isAdmin != null ? isAdmin : false);

        UserEntity saved = userJpaRepository.save(entity);
        logger.info("创建用户成功: userId={}, username={}", saved.getUserId(), username);

        return toDomainModel(saved);
    }

    /**
     * 根据邮箱获取用户
     *
     * 对应Python: def get_user_by_email(...)
     *
     * @param email 邮箱
     * @return 用户，不存在返回null
     */
    public User getUserByEmail(String email) {
        return userJpaRepository.findByEmail(email)
            .map(this::toDomainModel)
            .orElse(null);
    }

    /**
     * 根据手机号获取用户
     *
     * 对应Python: def get_user_by_phone(...)
     *
     * @param phone 手机号
     * @return 用户，不存在返回null
     */
    public User getUserByPhone(String phone) {
        return userJpaRepository.findByPhone(phone)
            .map(this::toDomainModel)
            .orElse(null);
    }

    /**
     * 根据用户名获取用户
     *
     * 对应Python: def get_user_by_username(...)
     *
     * @param username 用户名
     * @return 用户，不存在返回null
     */
    public User getUserByUsername(String username) {
        return userJpaRepository.findByUsername(username)
            .map(this::toDomainModel)
            .orElse(null);
    }

    /**
     * 根据用户ID获取用户
     *
     * 对应Python: def get_user_by_id(...)
     *
     * @param userId 用户ID
     * @return 用户，不存在返回null
     */
    public User getUserById(String userId) {
        return userJpaRepository.findById(userId)
            .map(this::toDomainModel)
            .orElse(null);
    }

    /**
     * 获取所有用户（分页）
     *
     * 对应Python: def get_all_users(...)
     *
     * @param page 页码（从1开始）
     * @param pageSize 每页数量
     * @param isAdmin 管理员筛选（可选）
     * @return 用户列表和总数
     */
    public UserListResult getAllUsers(int page, int pageSize, Boolean isAdmin) {
        // 限制每页最大数量
        if (pageSize > 100) {
            pageSize = 100;
        }

        // 创建分页和排序
        Pageable pageable = PageRequest.of(
            page - 1,  // Spring Data JPA页码从0开始
            pageSize,
            Sort.by(Sort.Direction.DESC, "createdAt")
        );

        Page<UserEntity> result;
        if (isAdmin != null) {
            result = userJpaRepository.findAllByIsAdmin(isAdmin, pageable);
        } else {
            result = userJpaRepository.findAll(pageable);
        }

        List<User> users = result.getContent().stream()
            .map(this::toDomainModel)
            .collect(Collectors.toList());

        return new UserListResult(users, result.getTotalElements());
    }

    /**
     * 更新用户信息
     *
     * 对应Python: def update_user(...)
     *
     * @param userId 用户ID
     * @param username 新用户名
     * @param nickname 新昵称
     * @param email 新邮箱
     * @param phone 新手机号
     * @param isAdmin 是否为管理员
     * @return 更新后的用户，失败返回null
     */
    @Transactional
    public User updateUser(
            String userId,
            String username,
            String nickname,
            String email,
            String phone,
            Boolean isAdmin
    ) {
        return userJpaRepository.findById(userId)
            .map(entity -> {
                // 检查用户名是否与其他用户冲突
                if (username != null && !username.equals(entity.getUsername())) {
                    if (userJpaRepository.existsByUsername(username)) {
                        throw new IllegalArgumentException("用户名已存在: " + username);
                    }
                    entity.setUsername(username);
                }

                // 检查邮箱是否与其他用户冲突
                if (email != null && !email.equals(entity.getEmail())) {
                    if (userJpaRepository.existsByEmail(email)) {
                        throw new IllegalArgumentException("邮箱已存在: " + email);
                    }
                    entity.setEmail(email);
                }

                // 检查手机号是否与其他用户冲突
                if (phone != null && !phone.equals(entity.getPhone())) {
                    if (userJpaRepository.existsByPhone(phone)) {
                        throw new IllegalArgumentException("手机号已存在: " + phone);
                    }
                    entity.setPhone(phone);
                }

                // 更新其他字段
                if (nickname != null) {
                    entity.setNickname(nickname);
                }
                if (isAdmin != null) {
                    // 防止取消最后一个管理员
                    if (!isAdmin && entity.getIsAdmin() && isLastAdmin(userId)) {
                        throw new IllegalArgumentException("不能取消最后一个管理员的管理员权限");
                    }
                    entity.setIsAdmin(isAdmin);
                }

                UserEntity saved = userJpaRepository.save(entity);
                logger.info("更新用户成功: userId={}", userId);

                return toDomainModel(saved);
            })
            .orElse(null);
    }

    /**
     * 删除用户
     *
     * 对应Python: def delete_user(...)
     *
     * @param userId 用户ID
     * @param currentUserId 当前用户ID
     * @return 是否删除成功
     */
    @Transactional
    public boolean deleteUser(String userId, String currentUserId) {
        // 不能删除自己
        if (userId.equals(currentUserId)) {
            throw new IllegalArgumentException("不能删除自己的账号");
        }

        return userJpaRepository.findById(userId)
            .map(entity -> {
                // 防止删除最后一个管理员
                if (entity.getIsAdmin() && isLastAdmin(userId)) {
                    throw new IllegalArgumentException("不能删除最后一个管理员");
                }

                userJpaRepository.delete(entity);
                logger.info("删除用户成功: userId={}", userId);
                return true;
            })
            .orElse(false);
    }

    /**
     * 重置用户密码
     *
     * 对应Python: def reset_password(...)
     *
     * @param userId 用户ID
     * @param newPassword 新密码
     * @return 是否重置成功
     */
    @Transactional
    public boolean resetPassword(String userId, String newPassword) {
        return userJpaRepository.findById(userId)
            .map(entity -> {
                entity.setPasswordHash(passwordEncoder.encode(newPassword));
                userJpaRepository.save(entity);
                logger.info("重置用户密码成功: userId={}", userId);
                return true;
            })
            .orElse(false);
    }

    /**
     * 验证密码
     *
     * @param userId 用户ID
     * @param password 密码
     * @return 是否正确
     */
    public boolean verifyPassword(String userId, String password) {
        return userJpaRepository.findById(userId)
            .map(entity -> passwordEncoder.matches(password, entity.getPasswordHash()))
            .orElse(false);
    }

    /**
     * 检查是否是最后一个管理员
     */
    private boolean isLastAdmin(String userId) {
        long adminCount = userJpaRepository.countByIsAdminTrue();
        if (adminCount <= 1) {
            return true;
        }

        // 如果还有其他管理员，检查当前用户是否是管理员
        return userJpaRepository.findById(userId)
            .map(UserEntity::getIsAdmin)
            .orElse(false);
    }

    /**
     * 将JPA实体转换为领域模型
     */
    private User toDomainModel(UserEntity entity) {
        User user = new User();
        user.setId(entity.getUserId());
        user.setUsername(entity.getUsername());
        user.setEmail(entity.getEmail());
        user.setIsAdmin(entity.getIsAdmin());
        user.setCreatedAt(entity.getCreatedAt());
        // nickname, phone, avatar, updatedAt 等字段在User领域模型中可能不存在
        // 这里根据实际的User领域模型进行调整
        return user;
    }

    /**
     * 用户列表结果（包含列表和总数）
     */
    public static class UserListResult {
        private final List<User> users;
        private final long total;

        public UserListResult(List<User> users, long total) {
            this.users = users;
            this.total = total;
        }

        public List<User> getUsers() {
            return users;
        }

        public long getTotal() {
            return total;
        }
    }
}
